ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/HD/BackingStore.hs

```diff
@@ -0,0 +1,373 @@
+import qualified Data.Set as Set
+import           Data.Typeable
+import qualified System.Directory as Dir
+import           System.IO.Temp (createTempDirectory)
+
+import           Cardano.Binary (FromCBOR (..), ToCBOR (..))
+
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck (Arbitrary (..), Property, Testable)
+import           Test.QuickCheck.Gen.Unsafe
+import qualified Test.QuickCheck.Monadic as QC
+import           Test.QuickCheck.Monadic (PropertyM)
+import           Test.Tasty
+import           Test.Tasty.HUnit
+import           Test.Tasty.QuickCheck (testProperty)
+
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Storage.FS.API hiding (Handle)
+import           Ouroboros.Consensus.Storage.FS.API.Types hiding (Handle)
+import           Ouroboros.Consensus.Storage.FS.IO (ioHasFS)
+import qualified Ouroboros.Consensus.Storage.LedgerDB.HD.BackingStore as BS
+import qualified Ouroboros.Consensus.Storage.LedgerDB.HD.DiffSeq as DS
+import qualified Ouroboros.Consensus.Storage.LedgerDB.HD.LMDB as LMDB
+import           Ouroboros.Consensus.Storage.LedgerDB.OnDisk
+import           Ouroboros.Consensus.Util ((.:))
+import           Ouroboros.Consensus.Util.IOLike hiding (MonadMask (..))
+
+import qualified Test.Util.FS.Sim.MockFS as MockFS
+import           Test.Util.FS.Sim.STM
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.Orphans.Slotting.Arbitrary ()
+import           Test.Util.Orphans.ToExpr ()
+import           Test.Util.TestLedgerState
+
+import           Test.QuickCheck.StateModel as StateModel
+import           Test.QuickCheck.StateModel.Lockstep as Lockstep
+import           Test.QuickCheck.StateModel.Lockstep.Run as Lockstep
+
+import           Test.Ouroboros.Storage.LedgerDB.HD.BackingStore.Lockstep
+import qualified Test.Ouroboros.Storage.LedgerDB.HD.BackingStore.Mock as Mock
+import           Test.Ouroboros.Storage.LedgerDB.HD.BackingStore.Registry
+
+{-------------------------------------------------------------------------------
+  Main test tree
+-------------------------------------------------------------------------------}
+
+-- TODO: use scaling instead of withMaxSuccess
+tests :: TestTree
+tests = testGroup "BackingStore" [
+    testProperty "InMemory IOSim SimHasFS" $ QC.withMaxSuccess 1000
+      testWithIOSim
+  , testProperty "InMemory IO SimHasFS" $ QC.withMaxSuccess 1000 $
+      testWithIO $
+        setupBSEnv InMemoryBackingStore setupSimHasFS (pure ())
+  , testProperty "InMemory IO IOHasFS" $ QC.withMaxSuccess 1000 $
+      testWithIO $ do
+        (fp, cleanup) <- setupTempDir
+        setupBSEnv InMemoryBackingStore (setupIOHasFS fp) cleanup
+  , testProperty "LMDB IO IOHasFS" $ QC.withMaxSuccess 200 $
+      testWithIO $ do
+        (fp, cleanup) <- setupTempDir
+        setupBSEnv (LMDBBackingStore testLMDBLimits) (setupIOHasFS fp) cleanup
+  , testCase "labelledExamples" labelledExamples
```

Comment:
Is it useful to run this on CI? I think it'd be better to perform coverage checking instead.

Reply:
I've removed it, since it will generate an unreasonable number of examples for even a small number of tags. This is because it tries to find an example *for all subsets of tags*, which means that if we have `n` possible tags, then it starts finding/printing a powerset `2^n` of examples/tags.